### PR TITLE
Support payment processing via platform checkout

### PR DESCRIPTION
**Pull request by reykjalin**
#### Changes proposed in this Pull Request

To enable payment processing via platform checkout we need to make some changes to the payment processing, mostly in the form of filters.

Related PRs are listed below via GitHub mentions. You must use the code in the related PRs for the changes in this PR to be effective.

#### Testing instructions

* Follow the testing instruction in PR 280 in the platform repo (linked below).

-------------------

- [ ] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : _Add link here / 'QA Testing Not Applicable'_

### CODE REVIEW

File: includes/class-wc-payment-gateway-wcpay.php [hunk 1]
```diff
@@ -2184,6 +2185,9 @@ public function update_order_status() {
 			}
 
 			$payment_method_id = isset( $_POST['payment_method_id'] ) ? wc_clean( wp_unslash( $_POST['payment_method_id'] ) ) : '';
+			if ( 'null' === $payment_method_id ) {
+				$payment_method_id = '';
+			}
```
Comment (reykjalin) [inline includes/class-wc-payment-gateway-wcpay.php:2190]: This was a weird little bug I ran into and I'm not exactly sure how it happens. But for whatever reason the value of `$_POST['payment_method_id']` was set to `"null"` (i.e. a string) when it was supposed to be `null`.

Honestly not sure how to reproduce, but adding the extra if-statement here fixed the issue.

File: includes/class-wc-payments.php [hunk 2]
```diff
@@ -789,16 +800,29 @@ public static function maybe_register_platform_checkout_hooks() {
 	public static function ajax_init_platform_checkout() {
 		$session_cookie_name = apply_filters( 'woocommerce_cookie', 'wp_woocommerce_session_' . COOKIEHASH );
 
+		$user        = wp_get_current_user();
+		$customer_id = self::$customer_service->get_customer_id_by_user_id( $user->ID );
+		if ( null === $customer_id ) {
+			// create customer.
+			$customer_data = WC_Payments_Customer_Service::map_customer_data( null, new WC_Customer( $user->ID ) );
+			self::$customer_service->create_customer_for_user( $user, $customer_data );
+		}
```
Comment (reykjalin) [inline includes/class-wc-payments.php:809]: This approach of creating the Customer like this before seems like the simplest way to get what we want:

- A customer ID we can pass to the platform checkout.
- A customer ID associated with the customer's current session — even if they're checking out as a guest.

Some alternative approaches that might be worth looking into;

- A re-usable function we can use throughout WCPay to initialize a customer.
- Creating the customer on the platform checkout side and just pass `null` here if the customer ID hasn't been created yet.

The current implementation work, so these are mostly notes that would be best to resolve and discuss in a separate issue.

**Comment by rafaelzaleski** [comment]
Hey @reykjalin, I know this PR is still a draft, but I just wanted to point out that I cannot save the settings on this branch. If you are aware of that, please ignore this comment.

Screen Shot 2022-02-04 at 10 28 12

**Comment by reykjalin** [comment]
@rafaelzaleski Thanks for looking into that! I'm not so sure this is caused by this PR; it may have been a problem in `develop` the last time I updated the branch.

I updated the branch just now and hopefully that will have fixed the issue.

In any case, I'm not going to spend time on debugging the issue right now because we want to try to make changes on the server side that will make it so that we won't need the changes in this PR. If we're not successful in completely eliminating the need for the changes in this PR I'll come back to this issue 🙂

**Comment by reykjalin** [comment]
Closed in favor of https://github.com/Automattic/woocommerce-payments/pull/3811

**Pull request closed without merging**